keps/sig-auth/4412-projected-service-account-tokens-for-kubelet-image-credential-providers/README.md

```diff
@@ -0,0 +1,1107 @@
+
+<!--
+
+Production readiness reviews are intended to ensure that features merging into
+Kubernetes are observable, scalable and supportable; can be safely operated in
+production environments, and can be disabled or rolled back in the event they
+cause increased failures in production. See more in the PRR KEP at
+https://git.k8s.io/enhancements/keps/sig-architecture/1194-prod-readiness.
+
+The production readiness review questionnaire must be completed and approved
+for the KEP to move to `implementable` status and be included in the release.
+
+In some cases, the questions below should also have answers in `kep.yaml`. This
+is to enable automation to verify the presence of the review, and to reduce review
+burden and latency.
+
+The KEP must have a approver from the
+[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
+team. Please reach out on the
+[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
+you need any help or guidance.
+-->
+
+### Feature Enablement and Rollback
+
+<!--
+This section must be completed when targeting alpha to a release.
+-->
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+<!--
+Pick one of these and delete the rest.
+
+Documentation is available on [feature gate lifecycle] and expectations, as
+well as the [existing list] of feature gates.
+
+[feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
+[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
+-->
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: `ServiceAccountTokenForKubeletCredentialProviders`
+  - Components depending on the feature gate: kubelet
+
+```go
+FeatureSpec{
+  Default: false,
+  LockToDefault: false,
+  PreRelease: featuregate.Alpha,
+}
+```
+
+###### Does enabling the feature change any default behavior?
+
+<!--
+Any change of default behavior may be surprising to users or break existing
+automations, so be extremely careful here.
+-->
+
+No. After enabling the feature flag in the kubelet, the credential provider still needs to opt into using service account tokens for image pull
+by setting the `TokenAttributes` field in the kubelet credential provider configuration.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+<!--
+Describe the consequences on existing workloads (e.g., if this is a runtime
+feature, can it break the existing applications?).
+
+Feature gates are typically disabled by setting the flag to `false` and
+restarting the component. No other changes should be necessary to disable the
+feature.
+
+NOTE: Also set `disable-supported` to `true` or `false` in `kep.yaml`.
+-->
+
+Yes. The feature flag needs to be disabled and the credential provider configuration for the provider that is using service account tokens for image pull
```

Comment:
What happens to the cache when the feature is disabled?  What about images that were already pulled?

Reply:
IIUC disabling the feature gate will result in kubelet on the node being restarted? I looked at https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ and didn't see a dynamic reload option, so I'm assuming it always restarts as part of disabling the feature. The cache is in-memory, so that'll be invalidated. For images that have already been pulled, the behavior will be same as today with any pod being able to access the image. With the ensure image pulls KEP, we will be able to force reauthentication if the image is stored in the `pulled` cache on disk. 

LMK what you think!

Reply:
Maybe mention that kubelet restart will remove the in-memory cache?

Reply:
done!